apps/cowswap-frontend/src/modules/permit/utils/generatePermitHook.ts

```diff
@@ -0,0 +1,101 @@
+import { MaxUint256 } from '@ethersproject/constants'
+
+import { Eip2612PermitUtils } from '@1inch/permit-signed-approvals-utils'
+
+import { GP_VAULT_RELAYER } from 'legacy/constants'
+
+import { Web3ProviderConnector } from './Web3ProviderConnector'
+
+import { FAKE_SIGNER } from '../const'
+import { PermitHookParams } from '../types'
+
+const cachePrefix = 'permitCache-'
+const pendingRequests: { [permitKey: string]: Promise<string> | undefined } = {}
+
+const FIVE_YEARS_IN_SECONDS = 5 * 365 * 24 * 60 * 60
+
+function getDeadline() {
+  return Math.ceil(Date.now() / 1000) + FIVE_YEARS_IN_SECONDS
+}
+
+export async function generateQuotePermitHook(params: PermitHookParams): Promise<string> {
+  const { inputToken, chainId, permitInfo, provider, account } = params
+  const tokenAddress = inputToken.address
+  const tokenName = inputToken.name || tokenAddress
+  const permitKey = `${cachePrefix}${inputToken.address.toLowerCase()}-${chainId}${account ? `-${account}` : ''}`
+
+  if (localStorage.getItem(permitKey)) return localStorage.getItem(permitKey)!
+  if (pendingRequests[permitKey]) return pendingRequests[permitKey]!
+
+  const web3ProviderConnector = new Web3ProviderConnector(provider, account ? undefined : FAKE_SIGNER)
```

Comment:
why not using allways the fake signer and catching it (since it doesn't change the address --> if u use a PK)

Reply:
Because some times (when sending the order for real) we want to use the provider to actually sign the permit which will be included the order.
But I can cache the fake version, true.

Reply:
The fake signer has been cached.

But I cannot cache the web3provider instance as it depends on the provider for fetching contract data